src/java.base/share/classes/jdk/internal/vm/Continuation.java

```diff
@@ -0,0 +1,525 @@
+ * questions.
+ */
+
+package jdk.internal.vm;
+
+import jdk.internal.misc.PreviewFeatures;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.internal.vm.annotation.IntrinsicCandidate;
+import sun.security.action.GetPropertyAction;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+/**
+ * A one-shot delimited continuation.
+ */
+public class Continuation {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final boolean PRESERVE_SCOPE_LOCAL_CACHE;
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+    static {
+        PreviewFeatures.ensureEnabled();
+
+        StackChunk.init(); // ensure StackChunk class is initialized
+
+        String value = GetPropertyAction.privilegedGetProperty("jdk.preserveScopeLocalCache");
+        PRESERVE_SCOPE_LOCAL_CACHE = (value == null) || Boolean.parseBoolean(value);
+    }
+
+    private static final VarHandle MOUNTED;
+
+    /** Reason for pinning */
+    public enum Pinned {
+        /** Native frame on stack */ NATIVE,
+        /** Monitor held */          MONITOR,
+        /** In critical section */   CRITICAL_SECTION }
+
+    /** Preemption attempt result */
+    public enum PreemptStatus {
+        /** Success */                                                      SUCCESS(null),
+        /** Permanent failure */                                            PERM_FAIL_UNSUPPORTED(null),
+        /** Permanent failure: continuation already yielding */             PERM_FAIL_YIELDING(null),
+        /** Permanent failure: continuation not mounted on the thread */    PERM_FAIL_NOT_MOUNTED(null),
+        /** Transient failure: continuation pinned due to a held CS */      TRANSIENT_FAIL_PINNED_CRITICAL_SECTION(Pinned.CRITICAL_SECTION),
+        /** Transient failure: continuation pinned due to native frame */   TRANSIENT_FAIL_PINNED_NATIVE(Pinned.NATIVE),
+        /** Transient failure: continuation pinned due to a held monitor */ TRANSIENT_FAIL_PINNED_MONITOR(Pinned.MONITOR);
+
+        final Pinned pinned;
+        private PreemptStatus(Pinned reason) { this.pinned = reason; }
+        /**
+         * Whether or not the continuation is pinned.
+         * @return whether or not the continuation is pinned
+         **/
+        public Pinned pinned() { return pinned; }
+    }
+
+    private static Pinned pinnedReason(int reason) {
+        switch (reason) {
+            case 2: return Pinned.CRITICAL_SECTION;
+            case 3: return Pinned.NATIVE;
+            case 4: return Pinned.MONITOR;
+            default:
+                throw new AssertionError("Unknown pinned reason: " + reason);
+        }
```

Comment:
```suggestion
        return switch (reason) {
            case 2 -> Pinned.CRITICAL_SECTION;
            case 3 -> Pinned.NATIVE;
            case 4 -> Pinned.MONITOR;
            default -> throw new AssertionError("Unknown pinned reason: " + reason);
        };
```

Reply:
That would be clear (I think this predates switch expressions).